File: packages/openbrowser/src/tools/automation.ts
import type { BrowserAutomationParams, ToolResult } from "../types.js"
import { BrowserManager } from "../browser/manager.js"

export async function browserAutomation(
  params: BrowserAutomationParams,
): Promise<ToolResult> {
  let browser
  let page
  const results: string[] = []
  const screenshots: string[] = []

  try {
    browser = await BrowserManager.getBrowser(params.profileId)
    page = await BrowserManager.createPage(browser, params.profileId)

    // Navigate to initial URL
    await page.goto(params.url, {
      waitUntil: "networkidle0",
      timeout: 30000,
    })

    results.push(`Navigated to ${params.url}`)

    // Execute each action
    for (const action of params.actions) {
      try {
        switch (action.type) {
          case "click":
            if (!action.selector) {
              throw new Error("Click action requires a selector")
            }
            await page.waitForSelector(action.selector, {
              timeout: action.timeout || 5000,
            })
            await page.click(action.selector)
            results.push(`Clicked on ${action.selector}`)
            break

          case "type":
            if (!action.selector || !action.text) {
              throw new Error("Type action requires selector and text")
            }
            await page.waitForSelector(action.selector, {
              timeout: action.timeout || 5000,
            })
            await page.type(action.selector, action.text)
            results.push(`Typed "${action.text}" into ${action.selector}`)
            break

          case "wait":
            if (action.selector) {
              await page.waitForSelector(action.selector, {
                timeout: action.timeout || 30000,
              })
              results.push(`Waited for ${action.selector}`)
            } else if (action.timeout) {
              await new Promise((resolve) =>
                setTimeout(resolve, action.timeout),
              )
              results.push(`Waited for ${action.timeout}ms`)
            } else {
              throw new Error("Wait action requires selector or timeout")
            }
            break

          case "scroll":
            if (action.selector) {
              await page.evaluate((selector) => {
                const element = document.querySelector(selector)
                if (element) {
                  element.scrollIntoView({
                    behavior: "smooth",
                    block: "center",
                  })
                }
              }, action.selector)
              results.push(`Scrolled to ${action.selector}`)
            } else {
              // Scroll to bottom
              await page.evaluate(() => {
                window.scrollTo(0, document.body.scrollHeight)
              })
              results.push("Scrolled to bottom of page")
            }
            await new Promise((resolve) => setTimeout(resolve, 1000)) // Wait for scroll to complete
            break

          case "screenshot":
            const screenshot = await page.screenshot({
              encoding: "base64",
              fullPage: action.selector ? false : true,
            })
            screenshots.push(screenshot)
            results.push("Captured screenshot")
            break

          default:
            results.push(`Unknown action type: ${action.type}`)
        }
      } catch (actionError) {
        const errorMsg =
          actionError instanceof Error
            ? actionError.message
            : String(actionError)
        results.push(`Error executing ${action.type}: ${errorMsg}`)
      }
    }

    // Get final page URL
    const finalUrl = page.url()

    // Build response
    const content: ToolResult["content"] = [
      {
        type: "text",
        text: `Browser automation completed:\n\n${results.join("\n")}\n\nFinal URL: ${finalUrl}`,
      },
    ]

    // Add screenshots
    for (const screenshot of screenshots) {
      content.push({
        type: "image",
        data: screenshot,
        mimeType: "image/png",
      })
    }

    return { content }
  } catch (error) {
    console.error("Browser automation error:", error)
    return {
      content: [
        {
          type: "text",
          text: `Error during browser automation: ${error instanceof Error ? error.message : String(error)}`,
        },
      ],
      isError: true,
    }
  } finally {
    if (page && !params.headless) {
      // Keep browser open for debugging if not headless
      console.log("Browser kept open for debugging")
    } else if (page) {
      await page.close().catch(() => {})
    }
  }
}

File: packages/tui/internal/browser/automation.go
package browser

import (
	"context"
	"fmt"
	"time"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/chromedp"
	"github.com/chromedp/chromedp/kb"
)

// BrowserAutomation performs automated browser actions
func (e *Engine) BrowserAutomation(params AutomationParams) (*AutomationResult, error) {
	if params.ProfileID == "" {
		params.ProfileID = "automation"
	}

	// Get or create context
	ctx, _ := e.getOrCreateContext(params.ProfileID)

	// Create timeout context
	timeoutCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
	defer cancel()

	result := &AutomationResult{
		Success: true,
		Actions: make([]ActionResult, 0),
	}

	// Navigate to initial URL if provided
	if params.URL != "" {
		if err := chromedp.Run(timeoutCtx, chromedp.Navigate(params.URL)); err != nil {
			return nil, fmt.Errorf("failed to navigate: %w", err)
		}
		result.FinalURL = params.URL

		// Check for CAPTCHA after navigation
		if err := e.DetectAndSolveCaptcha(timeoutCtx); err != nil {
			// Log CAPTCHA detection but don't fail
			result.Actions = append(result.Actions, ActionResult{
				Type:    "captcha_check",
				Success: false,
				Error:   err.Error(),
			})
		}
	}

	// Execute each action
	for _, action := range params.Actions {
		actionResult := e.executeAction(timeoutCtx, action)
		result.Actions = append(result.Actions, actionResult)

		if !actionResult.Success {
			result.Success = false
			result.Error = actionResult.Error
			break
		}
	}

	// Get final page content
	if result.Success {
		var content string
		var url string
		chromedp.Run(timeoutCtx,
			chromedp.Text("body", &content),
			chromedp.Location(&url),
		)
		result.FinalContent = content
		result.FinalURL = url
	}

	return result, nil
}

// executeAction executes a single browser action
func (e *Engine) executeAction(ctx context.Context, action Action) ActionResult {
	result := ActionResult{
		Type:    action.Type,
		Success: true,
	}

	var err error

	switch action.Type {
	case "click":
		// Try multiple strategies for clicking
		err = e.robustClick(ctx, action.Selector)
		if err == nil {
			result.Message = fmt.Sprintf("Clicked element: %s", action.Selector)
		}

	case "type":
		// Try multiple strategies for typing
		err = e.robustType(ctx, action.Selector, action.Text)
		if err == nil {
			result.Message = fmt.Sprintf("Typed text into: %s", action.Selector)
		}

	case "wait":
		if action.Selector != "" {
			err = chromedp.Run(ctx, chromedp.WaitVisible(action.Selector))
			if err == nil {
				result.Message = fmt.Sprintf("Waited for element: %s", action.Selector)
			}
		} else if action.Text != "" {
			// Parse duration - accept milliseconds as plain number
			var duration time.Duration
			if ms, parseErr := time.ParseDuration(action.Text + "ms"); parseErr == nil {
				duration = ms
			} else if d, parseErr := time.ParseDuration(action.Text); parseErr == nil {
				duration = d
			} else {
				duration = 1 * time.Second
			}
			time.Sleep(duration)
			result.Message = fmt.Sprintf("Waited for %v", duration)
		}

	case "screenshot":
		var screenshot []byte
		err = chromedp.Run(ctx, chromedp.FullScreenshot(&screenshot, 90))
		if err == nil {
			result.Screenshot = screenshot
		}

	case "scroll":
		if action.Selector != "" {
			err = chromedp.Run(ctx,
				chromedp.ScrollIntoView(action.Selector),
			)
		} else {
			// Scroll to bottom
			err = chromedp.Run(ctx,
				chromedp.Evaluate(`window.scrollTo(0, document.body.scrollHeight)`, nil),
			)
		}

	case "press":
		// Handle keyboard shortcuts
		key := action.Key
		if key == "" {
			key = action.Text
		}

		var keyCode string
		switch key {
		case "Enter":
			keyCode = kb.Enter
		case "Tab":
			keyCode = kb.Tab
		case "Escape":
			keyCode = kb.Escape
		case "Backspace":
			keyCode = kb.Backspace
		case "Delete":
			keyCode = kb.Delete
		case "ArrowUp":
			keyCode = kb.ArrowUp
		case "ArrowDown":
			keyCode = kb.ArrowDown
		case "ArrowLeft":
			keyCode = kb.ArrowLeft
		case "ArrowRight":
			keyCode = kb.ArrowRight
		default:
			err = fmt.Errorf("unknown key: %s", key)
		}

		if err == nil {
			err = chromedp.Run(ctx, chromedp.KeyEvent(keyCode))
		}
	case "select":
		// Select dropdown option
		err = chromedp.Run(ctx,
			chromedp.SetValue(action.Selector, action.Text),
		)

	case "navigate":
		err = chromedp.Run(ctx,
			chromedp.Navigate(action.Text),
		)

	default:
		err = fmt.Errorf("unknown action type: %s", action.Type)
	}

	if err != nil {
		result.Success = false
		result.Error = err.Error()
	}

	return result
}

// robustClick tries multiple strategies to click an element
func (e *Engine) robustClick(ctx context.Context, selector string) error {
	// Get current URL to determine site-specific selectors
	var currentURL string
	chromedp.Run(ctx, chromedp.Location(&currentURL))

	// Strategy 1: Standard click with wait
	err := chromedp.Run(ctx,
		chromedp.WaitVisible(selector),
		chromedp.Click(selector),
	)
	if err == nil {
		return nil
	}

	// Strategy 2: JavaScript click with scroll into view
	err = chromedp.Run(ctx,
		chromedp.Evaluate(fmt.Sprintf(`
			const elem = document.querySelector('%s');
			if (elem) {
				elem.scrollIntoView({behavior: 'smooth', block: 'center'});
				setTimeout(() => elem.click(), 100);
				true;
			} else {
				false;
			}
		`, selector), nil),
	)
	if err == nil {
		time.Sleep(200 * time.Millisecond) // Allow click to process
		return nil
	}

	// Strategy 3: Try with different selector strategies
	selectors := []string{
		selector,
		fmt.Sprintf("[aria-label*='%s']", selector),
		fmt.Sprintf("[data-testid='%s']", selector),
		fmt.Sprintf("[placeholder*='%s']", selector),
		fmt.Sprintf("[title*='%s']", selector),
		fmt.Sprintf("[alt*='%s']", selector),
	}

	for _, sel := range selectors {
		err = chromedp.Run(ctx,
			chromedp.Click(sel, chromedp.NodeVisible),
		)
		if err == nil {
			return nil
		}
	}

	// Strategy 4: Force click with JavaScript
	err = chromedp.Run(ctx,
		chromedp.Evaluate(fmt.Sprintf(`
			const elem = document.querySelector('%s');
			if (elem) {
				// Remove any overlays
				const overlays = document.querySelectorAll('[style*="z-index: 9"], [style*="z-index: 10"], .modal-backdrop, .overlay');
				overlays.forEach(o => o.style.display = 'none');
				
				// Force click
				const event = new MouseEvent('click', {
					view: window,
					bubbles: true,
					cancelable: true
				});
				elem.dispatchEvent(event);
				true;
			} else {
				false;
			}
		`, selector), nil),
	)
	if err == nil {
		return nil
	}

	// Strategy 5: Mouse click at element position
	var nodes []*cdp.Node
	err = chromedp.Run(ctx,
		chromedp.Nodes(selector, &nodes, chromedp.ByQuery),
	)
	if err == nil && len(nodes) > 0 {
		return chromedp.Run(ctx,
			chromedp.MouseClickNode(nodes[0]),
		)
	}

	return fmt.Errorf("failed to click element: %s", selector)
}

// robustType tries multiple strategies to type text
func (e *Engine) robustType(ctx context.Context, selector string, text string) error {
	// Strategy 1: Standard type
	err := chromedp.Run(ctx,
		chromedp.WaitVisible(selector),
		chromedp.Clear(selector),
		chromedp.SendKeys(selector, text),
	)
	if err == nil {
		return nil
	}

	// Strategy 2: JavaScript value setting
	err = chromedp.Run(ctx,
		chromedp.Evaluate(fmt.Sprintf(`
			const elem = document.querySelector('%s');
			if (elem) {
				elem.focus();
				elem.value = '%s';
				elem.dispatchEvent(new Event('input', {bubbles: true}));
				elem.dispatchEvent(new Event('change', {bubbles: true}));
				true;
			} else {
				false;
			}
		`, selector, text), nil),
	)
	if err == nil {
		return nil
	}

	// Strategy 3: Focus and type
	return chromedp.Run(ctx,
		chromedp.Focus(selector),
		chromedp.SendKeys(selector, text, chromedp.NodeVisible),
	)
}
